Userland/Applications/Screenshoter/ScreenshoterWidget.cpp

```diff
@@ -0,0 +1,181 @@
+            m_anchor_point = event.position();
+    };
+
+    virtual void mousemove_event(GUI::MouseEvent& event) override
+    {
+        if (m_anchor_point.has_value()) {
+            m_region = Gfx::IntRect::from_two_points(*m_anchor_point, event.position());
+            update();
+        }
+    };
+
+    virtual void mouseup_event(GUI::MouseEvent& event) override
+    {
+        if (event.button() == GUI::MouseButton::Left)
+            m_window->close();
+    };
+
+    virtual void paint_event(GUI::PaintEvent&) override
+    {
+        GUI::Painter painter(*this);
+        painter.clear_rect(m_window->rect(), Gfx::Color::Transparent);
+        painter.fill_rect(m_window->rect(), m_background_color);
+
+        if (m_region.is_empty())
+            return;
+
+        painter.clear_rect(m_region, Gfx::Color::Transparent);
+    }
+
+    virtual void keydown_event(GUI::KeyEvent& event) override
+    {
+        if (event.key() == Key_Escape) {
+            m_region = Gfx::IntRect();
+            m_window->close();
+        }
+    }
+
+    Optional<Gfx::IntPoint> m_anchor_point;
+    Gfx::IntRect m_region;
+    GUI::Window* m_window = nullptr;
+    Gfx::Color const m_background_color;
+};
+
+ScreenshoterWidget::ScreenshoterWidget()
+{
+    load_from_gml(screenshoter_window_gml);
+
+    m_whole_button = *find_descendant_of_type_named<GUI::RadioButton>("wholescreen");
+    m_active_button = *find_descendant_of_type_named<GUI::RadioButton>("activewindow");
+    m_custom_button = *find_descendant_of_type_named<GUI::RadioButton>("customregion");
+    m_delay_spinbox = *find_descendant_of_type_named<GUI::SpinBox>("delay");
+    m_copy_checkbox = *find_descendant_of_type_named<GUI::CheckBox>("copy_to_clipboard");
+    m_cancel_button = *find_descendant_of_type_named<GUI::Button>("cancel_button");
+    m_ok_button = *find_descendant_of_type_named<GUI::Button>("ok_button");
+
+    m_cancel_button->on_click = [&](auto) {
+        GUI::Application::the()->quit();
+    };
+
+    m_ok_button->on_click = [&](auto) {
+        GUI::Application::the()->active_window()->close();
+
+        String output_path = Core::DateTime::now().to_string(String::formatted("{}/screenshot-%Y-%m-%d-%H-%M-%S.png", Core::StandardPaths::home_directory()));
+
+        usleep(m_delay_spinbox->value() * 1000);
+
+        save_screenshot(output_path, m_custom_button->is_checked(), m_copy_checkbox->is_checked());
+    };
+}
+
+void ScreenshoterWidget::save_screenshot(String output_path, bool selected = false, bool copy_to_clipboard = false)
+{
+    dbgln("Taking screenshot");
+    Gfx::IntRect crop_region;
+    if (selected) {
+        auto window = GUI::Window::construct();
+        auto& container = window->set_main_widget<SelectableLayover>(window);
+
+        window->set_title("Screenshoter");
+        window->set_has_alpha_channel(true);
+        window->set_fullscreen(true);
+        window->show();
+        GUI::Application::the()->exec();
```

Comment:
It is not appropriate to re-enter `Application::exec()` like this when it's already on the stack.
Instead we should return from here and give back control to the event loop that way.
The window can be kept alive by making it global.